File: src/test/java/cz/jskrabal/proxy/config/ProxyConfigurationBasicTest.java
package cz.jskrabal.proxy.config;

import static org.junit.Assert.assertEquals;

import org.junit.Test;
import org.junit.runner.RunWith;

import cz.jskrabal.proxy.config.enums.ConfigurationParameter;
import cz.jskrabal.proxy.config.enums.IdGeneratorType;
import junitparams.JUnitParamsRunner;
import junitparams.Parameters;

/**
 * Created by janskrabal on 10/06/16.
 */
@RunWith(JUnitParamsRunner.class)
public class ProxyConfigurationBasicTest extends AbstractConfigurationTest {

	@Test
	public void initProxyConfigWithExistingValidJson() {
		addValueToConfig(ConfigurationParameter.NETWORK_PORT, HTTP_RESERVED_PORT);
		ProxyConfiguration proxyConfiguration = new ProxyConfiguration(config);

		int port = proxyConfiguration.getValue(ConfigurationParameter.NETWORK_PORT, Integer.class);
		assertEquals(HTTP_RESERVED_PORT, port);
	}

	@Test
	public void initProxyConfigWithExistingInvalidJson() {
		addValueToConfig(ConfigurationParameter.NETWORK_PORT, INVALID_STRING_VALUE);
		ProxyConfiguration proxyConfiguration = new ProxyConfiguration(config);

		int port = proxyConfiguration.getValue(ConfigurationParameter.NETWORK_PORT, Integer.class);
		assertEquals(ConfigurationParameter.NETWORK_PORT.getDefaultValue(), port);
	}

	@Test(expected = ClassCastException.class)
	public void getConfigValueTypeMismatchTest() {
		proxyConfig.getValue(ConfigurationParameter.NETWORK_PORT, String.class);
	}

	@SuppressWarnings("unused")
	private static Object[] provideDefaultValueTestParams() {
		return new Object[] {
				new Object[] { ConfigurationParameter.ID_GENERATOR },
				new Object[] { ConfigurationParameter.NETWORK_HOST },
				new Object[] { ConfigurationParameter.NETWORK_PORT }
		};
	}

	@Test
	@Parameters(method = "provideDefaultValueTestParams")
	public void testDefaultValue(ConfigurationParameter param) {
		assertDefaultValue(param);
	}

	@SuppressWarnings("unused")
	private static Object[] provideSetAndGetTestParams() {
		return new Object[] {
				new Object[] { ConfigurationParameter.ID_GENERATOR, IdGeneratorType.SEQUENCE.name() },
				new Object[] { ConfigurationParameter.NETWORK_HOST, LOCALHOST_IP },
				new Object[] { ConfigurationParameter.NETWORK_PORT, HTTP_RESERVED_PORT }
		};
	}

	@Test
	@Parameters(method = "provideSetAndGetTestParams")
	public void testSetAndGet(ConfigurationParameter param, Object value) {
		addValueToConfig(param, value);
		assertConfigValue(param, value);
	}

	@SuppressWarnings("unused")
	private static Object[] provideInvalidValueTestParams() {
		return new Object[] {
				new Object[] { ConfigurationParameter.ID_GENERATOR, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.NETWORK_HOST, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.NETWORK_PORT, INVALID_STRING_VALUE },
				new Object[] { ConfigurationParameter.NETWORK_PORT, INVALID_NUMBER_VALUE }
		};
	}

	@Test
	@Parameters(method = "provideInvalidValueTestParams")
	public void testInvalidValue(ConfigurationParameter param, Object invalidValue) {
		addValueToConfig(param, invalidValue);
		assertDefaultValue(param);
	}

}

File: src/test/kotlin/cz/jskrabal/proxy/config/ProxyConfigurationBasicTest.kt
package cz.jskrabal.proxy.config

import org.junit.Assert.assertEquals

import org.junit.Test
import org.junit.runner.RunWith

import cz.jskrabal.proxy.config.enums.ConfigurationParameter
import cz.jskrabal.proxy.config.enums.IdGeneratorType
import junitparams.JUnitParamsRunner
import junitparams.Parameters

/**
 * Created by janskrabal on 10/06/16.
 */
@RunWith(JUnitParamsRunner::class)
internal class ProxyConfigurationBasicTest : AbstractConfigurationTest() {

    @Test
    fun initProxyConfigWithExistingValidJson() {
        addValueToConfig(ConfigurationParameter.NETWORK_PORT, AbstractConfigurationTest.HTTP_RESERVED_PORT)
        val proxyConfiguration = ProxyConfiguration(config)

        val port = proxyConfiguration.getValue(ConfigurationParameter.NETWORK_PORT, Int::class.javaObjectType)
        assertEquals(AbstractConfigurationTest.HTTP_RESERVED_PORT.toLong(), port.toLong())
    }

    @Test
    fun initProxyConfigWithExistingInvalidJson() {
        addValueToConfig(ConfigurationParameter.NETWORK_PORT, AbstractConfigurationTest.INVALID_STRING_VALUE)
        val proxyConfiguration = ProxyConfiguration(config)

        val port = proxyConfiguration.getValue(ConfigurationParameter.NETWORK_PORT, Int::class.javaObjectType)
        assertEquals(ConfigurationParameter.NETWORK_PORT.defaultValue, port)
    }

    @Test(expected = ClassCastException::class)
    fun getConfigValueTypeMismatchTest() {
        proxyConfig.getValue(ConfigurationParameter.NETWORK_PORT, String::class.java)
    }

    private fun provideDefaultValueTestParams(): Array<Any> {
        return arrayOf(arrayOf<Any>(ConfigurationParameter.ID_GENERATOR), arrayOf<Any>(ConfigurationParameter.NETWORK_HOST), arrayOf<Any>(ConfigurationParameter.NETWORK_PORT))
    }

    @Test
    @Parameters(method = "provideDefaultValueTestParams")
    fun testDefaultValue(param: ConfigurationParameter) {
        assertDefaultValue(param)
    }

    private fun provideSetAndGetTestParams(): Array<Any> {
        return arrayOf(arrayOf(ConfigurationParameter.ID_GENERATOR, IdGeneratorType.SEQUENCE.name), arrayOf(ConfigurationParameter.NETWORK_HOST, AbstractConfigurationTest.LOCALHOST_IP), arrayOf(ConfigurationParameter.NETWORK_PORT, AbstractConfigurationTest.HTTP_RESERVED_PORT))
    }

    @Test
    @Parameters(method = "provideSetAndGetTestParams")
    fun testSetAndGet(param: ConfigurationParameter, value: Any) {
        addValueToConfig(param, value)
        assertConfigValue(param, value)
    }

    private fun provideInvalidValueTestParams(): Array<Any> {
        return arrayOf(arrayOf(ConfigurationParameter.ID_GENERATOR, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.NETWORK_HOST, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.NETWORK_PORT, AbstractConfigurationTest.INVALID_STRING_VALUE), arrayOf(ConfigurationParameter.NETWORK_PORT, AbstractConfigurationTest.INVALID_NUMBER_VALUE))
    }

    @Test
    @Parameters(method = "provideInvalidValueTestParams")
    fun testInvalidValue(param: ConfigurationParameter, invalidValue: Any) {
        addValueToConfig(param, invalidValue)
        assertDefaultValue(param)
    }

}
